handler/src/main/java/io/netty/handler/logging/LoggingHandler.java

```diff
@@ -320,19 +375,26 @@ protected String format(ChannelHandlerContext ctx, String eventName, Object firs
     /**
      * Generates the default log message of the specified event whose argument is a {@link ByteBuf}.
      */
-    private static String formatByteBuf(ChannelHandlerContext ctx, String eventName, ByteBuf msg) {
+    private static String formatByteBuf(ChannelHandlerContext ctx, String eventName, ByteBuf msg, boolean hexDump,
+                                        Charset charset) {
         String chStr = ctx.channel().toString();
         int length = msg.readableBytes();
         if (length == 0) {
             StringBuilder buf = new StringBuilder(chStr.length() + 1 + eventName.length() + 4);
             buf.append(chStr).append(' ').append(eventName).append(": 0B");
             return buf.toString();
         } else {
-            int rows = length / 16 + (length % 15 == 0? 0 : 1) + 4;
-            StringBuilder buf = new StringBuilder(chStr.length() + 1 + eventName.length() + 2 + 10 + 1 + 2 + rows * 80);
-
-            buf.append(chStr).append(' ').append(eventName).append(": ").append(length).append('B').append(NEWLINE);
-            appendPrettyHexDump(buf, msg);
+            StringBuilder buf;
+            if (hexDump) {
+                int rows = length / 16 + (length % 15 == 0? 0 : 1) + 4;
+                buf = new StringBuilder(chStr.length() + 1 + eventName.length() + 2 + 10 + 1 + 2 + rows * 80);
+                buf.append(chStr).append(' ').append(eventName).append(": ").append(length).append('B').append(NEWLINE);
+                appendPrettyHexDump(buf, msg);
+            } else {
+                buf = new StringBuilder(chStr.length() + 1 + eventName.length() + 2 + 10 + 1 + 2 + length);
+                buf.append(chStr).append(' ').append(eventName).append(": ").append(length).append('B').append(NEWLINE);
+                buf.append(msg.toString(charset));
```

Comment:
Honestly I think this will not really produce what you would expect many times as the ByteBuf may only hold partial data etc. I think what would be more useful would be to just call `msg.toString()` here which would log the capacity, readableBytes etc of the buffer.

Reply:
I'm not sure that is any useful. For example, when using `msg.toString()`, the log looks like this:

```
INFO: [id: 0xb64ad33b, L:/0:0:0:0:0:0:0:1:8443 - R:/0:0:0:0:0:0:0:1:61687] WRITE: DefaultFullHttpResponse(decodeResult: success, version: HTTP/1.1, content: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf(ridx: 0, widx: 607, cap: 1024))
HTTP/1.1 200 OK
content-type: text/plain
pragma: no-cache
content-length: 607
connection: keep-alive, 607B
UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf(ridx: 0, widx: 607, cap: 1024)
```

The content is completely hidden.

The original (hex dump enabled) log is this:

```
INFO: [id: 0xbffae09f, L:/0:0:0:0:0:0:0:1:8443 - R:/0:0:0:0:0:0:0:1:61755] READ: DefaultHttpRequest(decodeResult: success, version: HTTP/1.1)
GET / HTTP/1.1
Host: localhost:8443
User-Agent: curl/7.54.0
Accept: */*
Aug 09, 2018 1:24:14 PM io.netty.handler.logging.LoggingHandler write
INFO: [id: 0xbffae09f, L:/0:0:0:0:0:0:0:1:8443 - R:/0:0:0:0:0:0:0:1:61755] WRITE: DefaultFullHttpResponse(decodeResult: success, version: HTTP/1.1, content: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf(ridx: 0, widx: 607, cap: 1024))
HTTP/1.1 200 OK
content-type: text/plain
pragma: no-cache
content-length: 607
connection: keep-alive, 607B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 49 70 73 6f 20 66 61 63 74 6f 20 69 73 20 61 20 |Ipso facto is a |
|00000010| 4c 61 74 69 6e 20 70 68 72 61 73 65 2c 20 64 69 |Latin phrase, di|
|00000020| 72 65 63 74 6c 79 20 74 72 61 6e 73 6c 61 74 65 |rectly translate|
|00000030| 64 20 61 73 20 22 62 79 20 74 68 65 20 66 61 63 |d as "by the fac|
|00000040| 74 20 69 74 73 65 6c 66 22 2c 5b 31 5d 20 77 68 |t itself",[1] wh|
|00000050| 69 63 68 20 6d 65 61 6e 73 20 74 68 61 74 20 61 |ich means that a|
|00000060| 20 73 70 65 63 69 66 69 63 20 70 68 65 6e 6f 6d | specific phenom|
|00000070| 65 6e 6f 6e 20 69 73 20 61 20 64 69 72 65 63 74 |enon is a direct|
|00000080| 20 63 6f 6e 73 65 71 75 65 6e 63 65 2c 20 61 20 | consequence, a |
|00000090| 72 65 73 75 6c 74 61 6e 74 20 65 66 66 65 63 74 |resultant effect|
|000000a0| 2c 20 6f 66 20 74 68 65 20 61 63 74 69 6f 6e 20 |, of the action |
|000000b0| 69 6e 20 71 75 65 73 74 69 6f 6e 2c 20 69 6e 73 |in question, ins|
|000000c0| 74 65 61 64 20 6f 66 20 62 65 69 6e 67 20 62 72 |tead of being br|
|000000d0| 6f 75 67 68 74 20 61 62 6f 75 74 20 62 79 20 61 |ought about by a|
|000000e0| 20 70 72 65 76 69 6f 75 73 20 61 63 74 69 6f 6e | previous action|
|000000f0| 2e 20 49 74 20 69 73 20 61 20 74 65 72 6d 20 6f |. It is a term o|
|00000100| 66 20 61 72 74 20 75 73 65 64 20 69 6e 20 70 68 |f art used in ph|
|00000110| 69 6c 6f 73 6f 70 68 79 2c 20 6c 61 77 2c 20 61 |ilosophy, law, a|
|00000120| 6e 64 20 73 63 69 65 6e 63 65 2e 20 41 6e 20 65 |nd science. An e|
|00000130| 78 61 6d 70 6c 65 20 69 6e 20 6c 61 77 20 69 73 |xample in law is|
|00000140| 20 6d 6f 6e 65 79 20 6c 61 75 6e 64 65 72 69 6e | money launderin|
|00000150| 67 3a 20 74 68 65 20 61 63 74 20 69 73 20 6e 6f |g: the act is no|
|00000160| 74 20 69 70 73 6f 20 66 61 63 74 6f 20 69 6c 6c |t ipso facto ill|
|00000170| 65 67 61 6c 20 62 65 63 61 75 73 65 20 69 74 20 |egal because it |
|00000180| 69 73 20 61 6e 20 65 78 63 68 61 6e 67 65 20 62 |is an exchange b|
|00000190| 75 74 20 69 73 20 64 6f 6e 65 20 61 73 20 61 20 |ut is done as a |
|000001a0| 63 6f 76 65 72 20 66 6f 72 20 73 6f 6d 65 74 68 |cover for someth|
|000001b0| 69 6e 67 20 65 6c 73 65 2c 20 73 6f 20 74 68 65 |ing else, so the|
|000001c0| 20 61 63 74 20 70 75 74 73 20 74 68 65 20 61 63 | act puts the ac|
|000001d0| 74 69 6f 6e 73 20 6f 66 20 61 6e 20 69 6e 64 69 |tions of an indi|
|000001e0| 76 69 64 75 61 6c 20 69 6e 20 71 75 65 73 74 69 |vidual in questi|
|000001f0| 6f 6e 2e 20 41 20 63 6f 6d 6d 6f 6e 20 45 6e 67 |on. A common Eng|
|00000200| 6c 69 73 68 20 69 64 69 6f 6d 20 77 69 74 68 20 |lish idiom with |
|00000210| 61 20 73 69 6d 69 6c 61 72 20 6d 65 61 6e 69 6e |a similar meanin|
|00000220| 67 20 69 73 20 22 69 6e 20 61 6e 64 20 6f 66 20 |g is "in and of |
|00000230| 69 74 73 65 6c 66 22 2e 20 43 6f 6d 70 61 72 65 |itself". Compare|
|00000240| 20 61 6c 73 6f 20 22 62 79 20 69 74 73 65 6c 66 | also "by itself|
|00000250| 22 20 61 6e 64 20 22 70 65 72 20 73 65 22 2e    |" and "per se". |
+--------+-------------------------------------------------+----------------+
```

It shows the content, but it cuts the content to 16 bytes per line. The word "directly" straddles the second and the third line in the above example. If I were to search "directly" in this log, I would not be able to find it.

When using `msg.toString(charset)`, the log is this:

```
INFO: [id: 0x8e5db168, L:/0:0:0:0:0:0:0:1:8443 - R:/0:0:0:0:0:0:0:1:61824] WRITE: DefaultFullHttpResponse(decodeResult: success, version: HTTP/1.1, content: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf(ridx: 0, widx: 607, cap: 1024))
HTTP/1.1 200 OK
content-type: text/plain
pragma: no-cache
content-length: 607
connection: keep-alive, 607B
Ipso facto is a Latin phrase, directly translated as "by the fact itself",[1] which means that a specific phenomenon is a direct consequence, a resultant effect, of the action in question, instead of being brought about by a previous action. It is a term of art used in philosophy, law, and science. An example in law is money laundering: the act is not ipso facto illegal because it is an exchange but is done as a cover for something else, so the act puts the actions of an individual in question. A common English idiom with a similar meaning is "in and of itself". Compare also "by itself" and "per se".
```

I understand that a message may be fragmented in multiple `ByteBuf`s, but that is not what this PR tries to get around. I'm simply trying to have the `LoggingHandler` print what it would have printed, sans the hex dump format. Whatever is held in the `ByteBuf` that would have been hex dumped, would just print out as decoded by the specified `charset`.

This is useful when you have a specific string you want to search in your log. For example in our [code](https://github.com/google/nomulus/blob/master/java/google/registry/proxy/HttpsRelayProtocolModule.java#L75), the `LoggingHandler` is installed after an `HttpObjectAggregator`, so we know that it must be getting a full http response from the peer server. Without hex dump, we can quickly search the log and find the particular http responses that contain the content that we are looking for. This is not possible currently because the hex dump format does not preserve the original content.

Let me know if the point I'm trying to make is clear. Thanks.

Reply:
I am still not sure this is something we should include as it may not produce what a user would expect.... @ejona86 @Scottmitch @carl-mastrangelo @mosesn @bryce-anderson thoughts ?

Reply:
This is not changing the default behavior of any existing `LoggingHandler` constructors and I think if a user would like to have the raw output decoded using a specific charset the user probably knows what he wants.

Just showing hex dump without *any* pretty printing (that includes spaces between every two characters and the added newlines) in theory could work, albeit a bit convoluted (you have to convert your search string to hex first).

Maybe my use case is not as broad as a imagined. I think it is mostly useful when you know your payload is textual and you want do literal searches easily. I thought it would be a good addition to upstream for similar use cases. It's ultimately up to you guys if this is a good inclusion :-)

Reply:
sorry for the late response but I am not convinced that this is general helpful as you may have noticed. So I will close this for now as @ejona86 also seems to have some doubt.